Added ASCOM telescope client

gionkunz opened the pull request:
**THIS PR IS WORK IN PROGRESS. I'M TRYING TO GET SOME FEEDBACK**

### Description

A lot of amateur astronomers, including myself, would like to control their ASCOM telescope using Stellarium. Although there is a 3rd party tool called StellariumScope that enables us to use our ASCOM telescopes with Stellarium, it's a rather elegant solution. StellariumScope is swapping config files, creates a local server and connects Stellarium to that server. The whole solution is error prone and simply does not feel right.

This PR aims to implement ASCOM specific telescope controls into Stellarium in order to have native ASCOM support. Since ASCOM is only available on Windows, this feature is specifically for Windows users with the ASCOM Platform installed.

Fixes #406 

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?
- Tested using EQ ASCOM Simulator
- Tested using ASCOM.Simulator.Telescope

**Test Configuration**:
* Operating system: Windows 10

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [X] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

gionkunz commented:
CC @gzotti CC @alex-w 

Hi guys, this is my initial PR. It's still work in progress, but I'm trying to get some feedback from you guys.

As I stated in the issue #406 , my C++ skills are very very rusty and I first needed to get up to speed with the basics again. Also I faced some additional challenges with the nature of ASCOM and OLE COM early binding but more or less gained control over that.

**Where I need input**

1.  I need to somehow disable the compilation of the whole ASCOM telescope control and also remove the choice from the UI if compiled under Linux. There is the macro Q_OS_WIN, should I try to do some shimming of the ASCOM telescope or the ASCOMDevice class and just leave a compilable  stub implementation? Also I could use that marco to hide the ASCOM choice in the UI.
2. How do I run the tests? I wan't to write some basic tests at least but haven't figured out how to run them. Can you give me a hint?
3. Changes to TelescopeControl: I have added more capabilities to the TelescopeControl / TelescopClient base class. Specifically a Telescope can have the optional capability to stop the current slew action and it can also issue a sync command. I'm not sure how this is done with INDI or NexStar, but syncing and updating the telescope internal pointing model is quite common I guess. I have also added more short-cuts to stop the slew and sync telescopes: `<ctrl> + <alt> + num` will stop the slew command of the telescope if possible. `<ctrl> + <shift> + num` is issuing a sync. Those are changes to the existing implementation but I guess they are backwards compatible. Other client implementations will need to implement by the corresponding contributors if applicable. What do you think? With this new feature, StellariumScope users with have a smooth transition to native ASCOM support, since syncing and stopping was a great feature made possible with multiple local servers (ctrl + 1, ctrl + 3, ctrl + 4).
3. i18n: I have added a lot of new text and I have no clue how to handle translations. Maybe you can help mere here?

**What's still missing?**

- [ ] This PR does currently not compile on Linux and we need to disable certain parts completely and also update the UI to disable ASCOM on Linux.
- [ ] Implement some checks when on Windows but no ASCOM Platform is installed
- [ ] I18n needs to be implemented correctly
- [ ] Basic unit tests should be added

gzotti commented:
Sounds promising, thanks. I can hopefully try it on Sunday.

gzotti commented:
Hi!

I tried to build your branch now. However I faced several compiler warnings issues around L"String" vs. wchar_t* elements and C++11 and a few more.  

About your questions: 
1: Yes, you must disable all ASCOM features with `#ifdef Q_OS_WIN...#endif ` constructs. In the CMakeLists.txt there is a similar `IF(WIN32)...ENDIF(WIN32)` construct. For the UI file handling, you would have to hide any ASCOM specific buttons or list entries in a non-Windows build.
2: Tests: @alex-w ?
3: I never used StellariumScope, so I lack any experience. I hope your extensions can be easily implemented in the other protocols. 
4: See the other examples. Just use q_("My translatable text") instead of QString("My translatable text").  These texts get extracted and fed to Transifex for our community translators. No need to bother for you. 
5: Are these Jnow/J2000 and refraction queries a mandatory feature of all ASCOM implementations? If it is optional, then of course the settings must remain manual. 

I have not configured my git to send you a PR now and lack the magic commands. I could prepare a diff of my compiler warning fixes and first CmakeLists changes.

As said before: I have never used or even installed anything ASCOM. Now, I tried to configure an ASCOM telescope. Pressing connect leads to these last lines in log.txt before silently exiting:

> Initialization failed for device:  ""
> connectionType: 7  initString: "New Telescope 2:ASCOM:J2000:"
> Creating telescope "New Telescope 2:ASCOM:J2000:" ; name/type/equinox/params: "New Telescope 2" "ASCOM" J2000 ""
> TelescopeClientASCOM::TelescopeClientASCOM with telescope name  "New Telescope 2"  and ascomDeviceId  ""
> Initialization failed for device:  ""

Do I have to install any additional drivers?

alex-w commented:
2. `make test` or 'ctest' (important note: you should run cmake with option `-DENABLE_TESTING=1` for Stellarium)

gionkunz commented:
@gzotti, thanks for the review! I'll check those warnings and fix them ASAP. If you have done some work already, maybe you could just push it somewhere or post the diff if it was a lot of work for you. If not, I'm also okay re-doing it.

1: Perfect, I'll do that. 
4: Cool, will do!
5: The implementation is mandatory, but vendors can return EPOCH UNKNOWN. I think I will add a small display field to the ASCOM control panel which shows the EPOCH of the ASCOM scope and a checkbox which allows you to specify if you'd like to use the devices EPOCH or the manual setting. This way we can cover all use-cases I guess.

To you ASCOM questions. If you install the Platform you will have two Telescope Simulators available by default. Telescope Simulator and Telescope Simulator .NET. You need to Choose the scope you want to connect with using the ASCOM Chooser. Just click the "Choose ASCOM Telescope" Button to choose a simulator scope. I still need to add some prevention there if you try to connect without selecting a scope. Also, if you have installed Stellarium 64bit, you will have issues with the ASCOM OLE COM objects which are mostly 32 bit. The same Problem exists with automation software like Sequence Generator Pro. I still need to find a way to prevent and catching exceptions when connecting to a wrong driver.

gzotti commented:
Yes I can prepare a git diff or such with my changes. Hopefully tonight. 
Can you please provide download  links which ASCOM packages the user is supposed to install?
If there are issues on 64-bit Windows (which is now the by far most-downloaded version), should we limit the ASCOM support to 32-bit builds? Or else you would have to deal with these issues.

gionkunz commented:
Yeah, this is kind of an independent issue by itself. The ASCOM Platform is 32bit/64bit (any). The OLE COM components for ASCOM drivers can be developed for 32bit, 64bit and any. The problem is that if you're using many ASCOM components, like most astrophotographer do, you're ending up with at least 4 drivers (Telescope, Focuser, Filterwheel, Camera) or even more (rotator, flat box, dome, environment). In my experience, at least one of the drivers is 32bit only. This will force the user to install everything in 32bit. The autoguider application, image sequencer and planetarium needs to be 32bit to communicate with the 32bit only ASCOM driver. From my experience most astrophotographer are installing 32bit versions of all tools, since 32bit ASCOM drivers are always available, where 64bit drivers are sometimes not. However, I guess there are also users who are lucky enough that all of their ASCOM drivers are 64bit or any. In that case the whole application chain can be 64bit.

Given those assumptions, I think it would be the best option to catch this error at runtime and provide necessary guidance. If a 64bit telescope from a 32bit host was selected (vice versa), an error message should be displayed. This might force astrophotographer users of Stellarium to re-install the appropriate bit version of Stellarium if their scope drivers demand so. I think that's an appropriate compromise given the underlying problem or driver architecture incompatibility. Thoughts?

gzotti commented:
With this jungle it seems to really make sense to implement both, but you would be indeed allowed to fail with a clear-text warning panel explaining this issue to non-tech users. (Mentioning 32/64bit is OK. Just not "Error 3156" without explanation.) It is also possible to install both versions of Stellarum in parallel.  
It seems the ASCOM programs would be provided by the manufacturers? I have an EQ3 mount which works out of the box with current Stellarium (used EQ5/6 and EQ3 with versions 0.12 and later). From where do I get the ASCOM ecosystem install packages required to run it with ASCOM aware programs?

gzotti commented:
Hi again. See my patch here: [GZ_patch761.txt](https://github.com/Stellarium/stellarium/files/3622515/GZ_patch761.txt) You can use git-apply to activate those changes.

Note that I have not checked yet whether this excludes building on Linux completely, I guess something is missing or misconfigured. I am also not sure this is the prettiest solution for the wchar_t warnings.

gionkunz commented:
Hi there, thanks for the patch! I have an EQ6-R and I use EQASCOM which should also work with an EQ3 skyscan. Usually the recommendation is to connect directly to the RJ45 connector with a FTDI <-> USB adapter. However, it also works if your handcontroller has a USB connector.

gionkunz commented:
@gzotti, I have fixed the warnings with the help of your diff, thanks for that!

I have also added a feature to control the coordinate system as discussed. A user can choose whether he wants to rely on the information of the ASCOM device to determine coordinate system type or enforce the Stellarium general coordinate system settings of the scope (Equinox).

I'm installing a Ubuntu in my Hyper-V on windows to compile there and fix any remaining issues but my manual tests look good so far.

Now the only things remaining are:

- [ ] Final checks for compilation on non-windows systems and make sure no ASCOM / OLE related stuff creeps in there
- [ ] Implement some checks when on Windows but no ASCOM Platform is installed. Should be as simple as trying to connect an OLE Object from ASCOM (chooser) and if it fails, we know there's an issue and don't show the ASCOM option
- [ ] Handle the case when no device has been chosen, the scope is added and connected (without device). This causes an exception at the moment.
- [ ] Test the build out in the field for a whole night with a real mount and make sure it does not explode...

gzotti commented:
Hi! Sounds good, thanks.

I can hopefully test the build on an Ubuntu 19.04 notebook later today, or one evening next week. 
My main notebook will stay ASCOM free (can test these cases here), but I can probably field test ASCOM with my guiding netbook on my NEQ3 during an astro/Stellarium workshop I am giving in 2 weeks. But it will be first time ever for me to run anything with ASCOM, so I can even run a "n00b test" :-)

alex-w commented:
On Windows the plugin has issue in the GUI when ASCOM platform is not installed. I think the better choice here will be add the persistent GUI elements and manage enable/disable state for it on Windows.

gionkunz commented:
Thanks @alex-w for the feedback. I still need some time to test all the usecases since I will need a VM to test without ASCOM (dont want to uninstall ascom on my main computer). We already disable the majority of ASCOM features with the use of the ASCOMSupport class, but due to lack of testing its very likely that I have missed some parts. Will fix that soon.

gionkunz commented:
Hi @alex-w , after merging your PR, I now have errors in the files inside the common folder of the telescope plugin during build. Something with the CMakeLists.txt in there seems to be causing issues for me. The compiler / linker does not find the Qt includes for files inside the common folder. Do you have any idea what could cause that?

gionkunz commented:
I think the issue was the missing Qt::Core link target:
```
IF(WIN32)
add_library(TelescopeControl_ASCOM_common STATIC
    ASCOMSupport.hpp
    ASCOMSupport.cpp
    OLE.hpp
    OLE.cpp
    )
// I'VE ADDED THIS
target_link_libraries(TelescopeControl_ASCOM_common
    Qt5::Core
    )

target_link_libraries(TelescopeControl_common
    TelescopeControl_ASCOM_common
    )
ENDIF(WIN32)
```

alex-w commented:
Hmm... Let me fix it.

gionkunz commented:
@alex-w , no worries. I have already fixed it! :-) With your changes the CI /  CD was green the first time 🎉 :-)

gionkunz commented on plugins/TelescopeControl/CMakeLists.txt:
```diff
@@ -2,8 +2,8 @@
 SET(TELESCOPE_CONTROL_MAJOR "0")
-SET(TELESCOPE_CONTROL_MINOR "3")
-SET(TELESCOPE_CONTROL_PATCH "1")
+SET(TELESCOPE_CONTROL_MINOR "4")
```
@gzotti Is that something I should be doing? I though that this would be a good time to bump the minor version

gzotti replied:
Yes, seems reasonable to do. Please also follow the Warning to update resources/device_models.json. But @alex-w knows the plugin much better than I do.

gionkunz replied:
@alex-w can you elaborate on the device_models.json file? Looking at it, some of those devices also support ASCOM and then there's a huge list of devices which will additionally be supported by the direct support of ASCOM now. How should we handle this? Where is this file used?

alex-w replied:
Just update the version in device_models.json file - he is used for define supported devices for our own drivers of mounts.

gionkunz commented on plugins/TelescopeControl/src/TelescopeControl.cpp:
```diff
@@ -78,7 +78,7 @@ StelPluginInfo TelescopeControlStelPluginInterface::getPluginInfo() const
 	info.id = "TelescopeControl";
 	info.displayedName = N_("Telescope Control");
-	info.authors = "Bogdan Marinov, Johannes Gajdosik, Alessandro Siniscalchi";
+	info.authors = "Bogdan Marinov, Johannes Gajdosik, Alessandro Siniscalchi, Gion Kunz";
```
@gzotti I added myself here, hope thats okay?

gzotti replied:
Definitely yes!

gionkunz commented on plugins/TelescopeControl/src/TelescopeControl.cpp:
```diff
@@ -180,9 +185,25 @@ void TelescopeControl::init()
+			shortcut = QString("Ctrl+Shift+Alt+%1").arg(i);
+			text = q_("Abort last slew command of telescope #%1").arg(i);
+			StelAction* actionAbortSlew = addAction(name, section, text, abortSlew, "map()", shortcut);
+			abortSlew->setMapping(actionAbortSlew, i);
```
@gzotti Should we document the new keyboard shortcuts somewhere?

alex-w replied:
Yes, in the Stellarium User Guide and Help window of plugin

gzotti replied:
I hope you are already writing an ASCOM section in the Stellarium User Guide (/guide/plg_atTheTelescope.tex). If the new hotkeys only apply to ASCOM telescopes, it should go there, else into the general plugin description. The hotkey actions themselves should be automatically listed in the F1 help screen. 

Speaking of User Guide: What do we do with StellariumScope? Is it obsolete now, or are there still scenarios for its use? (Networked solutions?) I am just not enough of an observer currently to see all the various applications and variants of using Stellarium with a gazillion of telescope mounts. 

While I see the "map()" function called: I have recently seen that the QSignalMapper is deprecated/obsolete, and we should rather get rid of it. In some Qt forum I found descriptions how to update C++ code with lambda functions etc, which is fragile territory for me. Not sure if we can ask you to do that for the telescope plugin, or @alex-w  do you know already what to do? They are used in TelescopeControl and Ocular plugins, those are the ones I currently use the least, personally, so I may miss something when I just change and make a few tests.

alex-w replied:
@gzotti I’ve tried use lambda for change the QSignalMapper but without success, because lambda’s cannot use QString. I didn’t make a “wrapper” for it.

gzotti replied:
@alex-w  Ouch. Is it possible to use QString.toByteArray().data() or something like that?

gionkunz replied:
@gzotti I think StellariumScope is obsolete now. Since I have implemented the new telescope control sync and abort slew commands, all the additional functionality which was provided by StellariumScope is covered.

alex-w replied:
> @alex-w Ouch. Is it possible to use QString.toByteArray().data() or something like that?

I don’t know because I didn’t touch lambdas after previous experiment

gionkunz commented:
I have tested the latest release build on a vanilla Windows 10 developer VM with no ASCOM and the GUI behaves as expected (no ASCOM options or errors) so we should be good there. I think we are getting closer!
![windows-no-ascom-smaller](https://user-images.githubusercontent.com/2203704/66567291-28ca0d80-eb67-11e9-9f9f-0026f7e7b1c8.gif)

alex-w commented:
@gionkunz I've added couple improvements - please check the PR

alex-w commented:
OK,  CI/CD is green now...

gionkunz commented:
@alex-w working on the tests right now. I'm struggling a bit with ctest. Can you show me an example command which allows me to build and run the tests?

alex-w commented:
@gionkunz are you using IDE (QtCreator or MSVC) or CLI?

gionkunz commented:
> @gionkunz are you using IDE (QtCreator or MSVC) or CLI?

I'm using Visual Studio 2017 but I can't see any tests in the test explorer. I've also tried from the command line with ctest but can't figure out the necessary input parameters or preparation steps.

alex-w commented:
I guess you are configure project without testing support. Please run cmake and enable option ENABLE_TESTING

> 1. `make test` or `ctest` (important note: you should run cmake with option `-DENABLE_TESTING=1` for Stellarium)

gionkunz commented:
> I guess you are configure project without testing support. Please run cmake and enable option ENABLE_TESTING
> 
> > 1. `make test` or `ctest` (important note: you should run cmake with option `-DENABLE_TESTING=1` for Stellarium)

Thanks, but how does that translate to Visual Studio, do you know? Can I add this to the CMakeSettings.json file?
![image](https://user-images.githubusercontent.com/2203704/66651424-406ec800-ec33-11e9-9173-228e2821243b.png)

alex-w commented:
Hmm... How you configure project? Are you using cmake GUI and VS or are you using cmake plugin within VS?

gionkunz commented:
@alex-w I rebuilt the CMake cache with those settings in VS 2017 inside that json file and now I see the tests in the test explorer :-) Thanks!

gionkunz commented:
Well, that was too early... :-) I see the tests but if I execute them (does not matter which one) they are failing. I'm looking into it but so far I can't see what's causing that.

![image](https://user-images.githubusercontent.com/2203704/66662102-7bc7c180-ec48-11e9-9b3c-8e805c9b6e1c.png)

alex-w commented:
Please run buildTest solution first.

gionkunz commented:
Hi @alex-w 

I'm making slow progress here, but slow is better than no :-P

I can run the tests now from VS2017 by using the CMake plugin menu CMake -> Tests -> Run Stellarium-CTests.

```
Test project C:/Users/gionk/CMakeBuilds/a465e9e1-6164-9234-84bd-e69845f73c6a/build/x64-Debug
      Start  1: testDates
 1/25 Test  #1: testDates ........................   Passed    3.35 sec
      Start  2: testStelFileMgr
 2/25 Test  #2: testStelFileMgr ..................   Passed    0.12 sec
      Start  3: testStelSphereGeometry
 3/25 Test  #3: testStelSphereGeometry ...........   Passed    1.94 sec
      Start  4: testStelSphericalIndex
 4/25 Test  #4: testStelSphericalIndex ...........   Passed    0.57 sec
      Start  5: testStelJsonParser
 5/25 Test  #5: testStelJsonParser ...............   Passed    0.34 sec
      Start  6: testStelIniParser
 6/25 Test  #6: testStelIniParser ................   Passed    0.07 sec
      Start  7: testStelVertexArray
 7/25 Test  #7: testStelVertexArray ..............   Passed    1.73 sec
      Start  8: testDeltaT
 8/25 Test  #8: testDeltaT .......................   Passed    0.12 sec
      Start  9: testConversions
 9/25 Test  #9: testConversions ..................   Passed    0.12 sec
      Start 10: testComputations
10/25 Test #10: testComputations .................   Passed    0.12 sec
      Start 11: testExtinction
11/25 Test #11: testExtinction ...................   Passed    0.03 sec
      Start 12: testRefraction
12/25 Test #12: testRefraction ...................   Passed    0.03 sec
      Start 13: testAirmass
13/25 Test #13: testAirmass ......................   Passed    0.03 sec
      Start 14: testGeomMath
14/25 Test #14: testGeomMath .....................   Passed    0.03 sec
      Start 15: testVecMath
15/25 Test #15: testVecMath ......................   Passed    0.03 sec
      Start 16: testStelProjector
16/25 Test #16: testStelProjector ................   Passed    0.03 sec
      Start 17: testPrecession
17/25 Test #17: testPrecession ...................   Passed    0.03 sec
      Start 18: testEphemeris
18/25 Test #18: testEphemeris ....................   Passed    0.54 sec
      Start 19: testStelSkyCultureMgr
19/25 Test #19: testStelSkyCultureMgr ............   Passed    0.15 sec
      Start 20: testNavStars
20/25 Test #20: testNavStars .....................   Passed    0.03 sec
      Start 21: testHistoricalSupernovae
21/25 Test #21: testHistoricalSupernovae .........   Passed    0.03 sec
      Start 22: testTelescopeControl_INDI
22/25 Test #22: testTelescopeControl_INDI ........   Passed    0.06 sec
      Start 23: testINDIConnection
23/25 Test #23: testINDIConnection ...............***Exception: Exit code 0xc0000135
  0.02 sec
      Start 24: testTelescopeClientINDI
24/25 Test #24: testTelescopeClientINDI ..........***Exception: Exit code 0xc0000135
  0.06 sec
      Start 25: testTelescopeControl_ASCOM
25/25 Test #25: testTelescopeControl_ASCOM .......***Exception: Exit code 0xc0000135
  0.12 sec

88% tests passed, 3 tests failed out of 25

Total Test time (real) =  10.14 sec

The following tests FAILED:
	 23 - testINDIConnection (Exit code 0xc0000135
)
	 24 - testTelescopeClientINDI (Exit code 0xc0000135
)
	 25 - testTelescopeControl_ASCOM (Exit code 0xc0000135
)
Errors while running CTest
```

However, it's weird because for me the testTelescopeClientINDI was failing, even if I have not changes anything in INDI, except the new sync and cancel virtual methods I've added to the abstract super-class. I tried to isolate the issue and found the problem for me was that testTelescopeClientINDI and testINDIConnection have the following include `#include <QTest>` but all other working tests I compared with have a different include `#include <QtTest>`. If I change the INDI tests to use the QtTest include instead, they start working.

Can you clarify on that?

alex-w commented:
Please change include to QtTest

gionkunz commented:
Okay, the weirdness continues. I still have failing tests but it's really weird. testTelescopeControl_INDI is running fine. So I have striped everything from my failing test testTelescopeClientINDI and now it looks exactly the same like testTelescopeControl_INDI. Empty class and no test methods. Still the test fails...

If I run the test executable with a regular debug run, testTelescopeControl_INDI.exe runs fine. When I run testTelescopeClientINDI.exe I get the runtime error telling me that the library SpoutLibrary.dll could not be found. This is totally weird and specially since I have rewritten the test to look exactly the same as testTelescopeControl_INDI, so basically doing nothing.

Do you have any clues? It's really hard for me to provide tests and I'm running in circles here.

alex-w commented:
See for example: https://ci.appveyor.com/project/alex-w/stellarium/builds/28064917/job/sd65r2q1c13lin98

Probably not all environment variable was set correctly in your development environment.

gionkunz commented:
That's really really weird. My CMake output was also telling me that SpoutLibrary.dll was found. However, when I run the test, I get this error. At least for specific tests but not for others. Now, I have manually added "C:/stellarium/util/spout2/x64" to my path and all tests are green :-/

gionkunz commented:
I'm using MSVC 2017 cmake instead of the one you recommended to install, that is maybe the reason for all this weird behavior.

gzotti commented:
I have finally built and am testing on my obs. netbook. (Win10, AMD A4-1200 APU, 1GHz. Not your favorite dev machine...). First observations from a total ASCOM beginner:
1) Without ASCOM platform, no sign of ASCOM. Good. 
2) Installed ASCOMPlatform64.exe. With only ASCOM installed, when I want to select the telescope, the ASCOM chooser panel hides behind the Stellarium full-screen app. This may be similar to https://github.com/Stellarium/stellarium/wiki/Common-Problems-for-the-current-version#some-system-dialogs-colors-screenshot-dir-hidden, but the color dialogs are visible. A warning on this should go into the User Guide then. 
3) I then selected the telescope simulator. Wanted to slew, but an ASCOM warning told me I have to unpark first. I tried to call the ASCOM telescope chooser dialog again, but Stellarium with ASCOM chooser panel on top froze. I could switch (Alt-Tab) to other programs, but cannot switch back to Stellarium. It is still in the background but apparently in some wait loop (?). Will start over, but only after upgrading to latest commits.

The ASCOM manual (current version packed into their installer is from 2011) states that ASCOM compliant software (that can show the logo) must run on XP, Vista and Win7. Is this still valid when XP and Vista are obsolete? I hope current demands would ask for 7(until 1/2020?)/8.*/10 only. 

Congrats for the international response so far!

gzotti commented:
OK, recompiled and restarted Stellarium. Again, with the "simulator" telescope. The ASCOM chooser no longer hides, this may have been a "first time" issue (?). However, it seems Stellarium or ASCOM hangs when I want to change telescope settings while Stellarium is connected to the ASCOM simulator. I had to kill Stellarium with task manager. On the next attempt ASCOM complained about config "panel already shown, cannot show modally". Only after reboot, I can config the telescope again, so maybe some initialisation could avoid that? 

Hmm, it seems that even after having connected and disconnected, calling the "Properties" button in the ASCOM chooser causes Stellarium to freeze again. Is this just with the simulator, or also with real telecopes? 

Also, the telescope config panel has fields for location data. Is it possible to read&transfer these data from the location set in Stellarium, or, vice versa, take coordinates from the telescope (may be GPS based) and set into Stellarium?

gzotti commented on plugins/TelescopeControl/src/gui/TelescopeDialog.cpp:
```diff
@@ -260,9 +263,10 @@ void TelescopeDialog::setAboutText()
+	aboutPage += "<tr><td rowspan=4><strong>" + q_("Authors") + "</strong></td><td>Johannes Gajdosik</td></td>";
 	aboutPage += "<tr><td>Michael Heinz</td></tr>";
 	aboutPage += "<tr><td>Bogdan Marinov &lt;bogdan.marinov84@gmail.com&gt; (" + q_("Plug-in and GUI programming") + ")</td></tr>";
+	aboutPage += "<tr><td>Gion Kunz &lt;gion.kunz@gmail.com&gt; (" + q_("ASCOM Telescope Client") + ")</td></tr>";
```
If we still discern between authors (original plugin designers) and contributors (later modifiers): Given that Petr added RTS2, Alessandro added INDI, this should make you a "contributor" of ASCOM support, please.

gzotti commented:
Sorry, but I must insist on a section for the User Guide for total beginners before giving my OK: A few words on what is ASCOM, where to download, and how to configure. How to deal with the ASCOM telescope coordinates vs. Stellarium's observer coordinates? (add bi-directional setting option?)  Also: 

It seems StellariumScope was a mechanism to deal also with networked telescopes (IP address field). Does some ASCOM-based setup provide the same already, making S.S. really obsolete, or will this be only available via Alpaca, and does Alpaca require further changes? I never had such hardware, so I simply don't know this. But there will be questions without doubt.

I still don't know why Stellarium freezes when I call the Properties button (for the simulator) a second time. It seems some observers on CloudyNights had more success with their real telescopes.

gionkunz commented:
@gzotti Absolutely nothing to feel sorry about! :-) I'm working on the documentation but didn't had time to work on this the past few days. I will for sure first add the necessary documentation before we go any further.

Stellarium Scope is starting a server and is altering the local Stellarium configuration and adds a remote telescope configuration (actually 3, slew, sync and abort) which connect to that local server(s). Then Stellarium Scope is acting as a broker between ASCOM and the remote telescope in Stellarium.

Technically it is possible to run Stellarium Scope on the Windows Machine where the Telescope is connected and then remote connect to it using Stellarium. Actually Stellarium Scope was never really made for this. The fact that it starts Stellarium Locally and modifies the configuration of the locally installed Stellarium proves that. They just use TCP / IP as a means of decoupling applications, system independently. One could however, if the technical knowledge allows it, copy or create the configuration for the 3 telescopes (slew, sync and cancel) on a remote Stellarium installation and point the IP address in Stellarium remote telescope control to the machine where Stellarium Scope is running.

This is definitely not a use-case for Stellarium Scope and the Developer even explains on the Stellarium Scope website that TCP / IP is used but not for remote connecting but local connection.

With all this in mind, I'd suggest we consider Stellarium Scope to be obsolete. Supporting remote ASCOM using Alpaca Client to Windows machines which use Remote ASCOM Alpaca server or to any other devices that runs an Alpaca compliant server, is something we should consider in the future. This will possible even be a different Telescope Client implementation (ASCOM Alpaca) since it's all based on HTTP and the implementation will differ a lot.

As far as I understood, the Alpaca protocol currently relies on a manual step, which is to figure out the device ID on the server, before you can connect to the telescope. They are working on a discovery service, which will allow you to query for available devices, but I guess this is still all work in progress. Once they support that officially and it's documented in their Swager doc, we should consider added an other telescope control for ASCOM Alpaca.

gzotti commented:
Thanks a lot for pointing this out. That means, any remote/networked solution was more "possible hack" than "designed use". Then I agree to declare this option obsolete. The S.S. section in the SUG can start with an obsolescence note, and the section removed in the SUG for version 0.20.0.  (For those who did it before, I think nothing actually changes, and they can continue to use S.S. and just follow older user guides?). 

Regarding Alpaca, if this is not ready for prime time, we can postpone this of course. But you likely know better. You are cordially invited to expand on this later :-)

If I understand correctly, "ASCOM compliant" still requires XP and Vista support. Or is there a statement from the ASCOM developers that remedies this statement?

gzotti commented:
Just tested on another, stronger notebook. Now I can modify telescope settings, e.g. slew speed, and have to re-connect after this modification. (This is OK, though, as this selection is usually a one-time thing setting). 
However, when I exit Stellarium while connected to the ASCOM simulator telescope, Stellarium crashes.  When I disconnect first, Stellarium exits normally. The logfile has no hint. Bug in the simulator, or do real telescopes behave similarly?

gionkunz commented:
Did you already built with my latest commits where I added the correct COM
disposal?

Georg Zotti <notifications@github.com> schrieb am Di., 15. Okt. 2019, 22:49:

> Just tested on another, stronger notebook. Now I can modify telescope
> settings, e.g. slew speed, and have to re-connect after this modification.
> (This is OK, though, as this selection is usually a one-time thing setting).
> However, when I exit Stellarium while connected to the ASCOM simulator
> telescope, Stellarium crashes. When I disconnect first, Stellarium exits
> normally. The logfile has no hint. Bug in the simulator, or do real
> telescopes behave similarly?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Stellarium/stellarium/pull/761?email_source=notifications&email_token=AAQ2AOD5NKAILJZXMKHXWW3QOYUFZA5CNFSM4IVZTKZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBKFPNA#issuecomment-542398388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ2AOGWRY7XYMYOSH4TVATQOYUFZANCNFSM4IVZTKZQ>
> .
>

gzotti commented:
Yes, commit is "Added some tests...". 
Notebook is a i7-640M, which runs circles around the 4 years newer netbook, so I think this is more what people use in the field now. (However, battery lasts longer on the AMD.)

gzotti commented:
Going back to the netbook, closing Stellarium while simulated telescope is connected works fine. 
I installed EQASCOM and configured a non-existent EQ5/6. Trying to connect..., and during this time shutdown also works, but leaves an ASCOM panel for the telescope open which I cannot close. On restart of Stellarium the panel vanished.

gzotti commented:
I finally was able to test under the sky. NEQ3 with ASCOM Skywatcher driver 6.0.6262.15273. Works nicely, but again, Stellarium crashed on closing while I was still connected to the telescope. Maybe there is some timing issue when disabling/freeing resources? I later tried several times again with just the simulator. Sometimes it works, sometimes it crashes when connected and closing the program. 

The telescope config panel again shows ways to set geographical location.  Is this standard for all ASCOM telescopes? Then I would really like to see a bidirectional config option in the main Stellarium ASCOM GUI to either set telescope location from Stellarium, or set Stellarium's location from what is set in the telescope.

gionkunz commented:
Thanks for all your testing work! Ill soon habe time again to work on this
project and use your feedback to do more tests.

Georg Zotti <notifications@github.com> schrieb am Mo., 21. Okt. 2019, 01:19:

> I finally was able to test under the sky. NEQ3 with ASCOM Skywatcher
> driver 6.0.6262.15273. Works nicely, but again, Stellarium crashed on
> closing while I was still connected to the telescope. Maybe there is some
> timing issue when disabling/freeing resources? I later tried several times
> again with just the simulator. Sometimes it works, sometimes it crashes
> when connected and closing the program.
>
> The telescope config panel again shows ways to set geographical location.
> Is this standard for all ASCOM telescopes? Then I would really like to see
> a bidirectional config option in the main Stellarium ASCOM GUI to either
> set telescope location from Stellarium, or set Stellarium's location from
> what is set in the telescope.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Stellarium/stellarium/pull/761?email_source=notifications&email_token=AAQ2AOC2XX5YVDJDMZ2HBMLQPTRPXA5CNFSM4IVZTKZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBYWQJQ#issuecomment-544303142>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ2AOGWULE2QJUWGSJHQ33QPTRPXANCNFSM4IVZTKZQ>
> .
>

gzotti commented:
One more question by that earlier ASCOM ignorant (me). "Sync to object" is something that AFAICS our classical control protocols don't support, we always said Stellarium only sends out GOTO commands and users must polar align correctly first. Is it mandatory for ASCOM? Then this is a clear advantage and justifies me switching over :-)  For my test, I had set my telescope not very accurately, aligned with 3 stars (result was a few degrees off correct polar alignment), and the drive managed to go to objects that were on the edge of my wide-angle eyepiece. (I know how to improve alignment for photography, but this is not the point here.) On the Stellarium screen the telescope marker also was a bit off the center of the object, but I did not absolutely check the pointing accuracy. When I moved the object into the center of the eyepiece and did "Sync to object", the marker went to the object. However, on next Slew to (next) object, the target was again on the edge of the eyepiece FoV, and marker was (correctly) not on the object. 
I should repeat the tests however, as the ASCOM driver told me I had only a version 2 handcontroller, some things may not work correctly, and I should upgrade to V 3 or 4. That HC says "version 3.35" which should fall into the "V3" classification (?). Not sure if this is responsible for sync not helping to raise accuracy for the next goto.

iriise commented:
Hi, I can confirm this! If I goto a star with the AstroArt Atlas and then solve and sync the image the star is centered in my image but the scope field in Stellarium shows the star near the edge of the scope circle! If I then do a goto from Stellarium the star moves to the edge of my image. I have J now set in AA and Stellarium.
> 
> One more question by that earlier ASCOM ignorant (me). "Sync to object" is something that AFAICS our classical control protocols don't support, we always said Stellarium only sends out GOTO commands and users must polar align correctly first. Is it mandatory for ASCOM? Then this is a clear advantage and justifies me switching over :-) For my test, I had set my telescope not very accurately, aligned with 3 stars (result was a few degrees off correct polar alignment), and the drive managed to go to objects that were on the edge of my wide-angle eyepiece. (I know how to improve alignment for photography, but this is not the point here.) On the Stellarium screen the telescope marker also was a bit off the center of the object, but I did not absolutely check the pointing accuracy. When I moved the object into the center of the eyepiece and did "Sync to object", the marker went to the object. However, on next Slew to (next) object, the target was again on the edge of the eyepiece FoV, and marker was (correctly) not on the object.
> I should repeat the tests however, as the ASCOM driver told me I had only a version 2 handcontroller, some things may not work correctly, and I should upgrade to V 3 or 4. That HC says "version 3.35" which should fall into the "V3" classification (?). Not sure if this is responsible for sync not helping to raise accuracy for the next goto.

gionkunz commented:
> Thanks a lot for pointing this out. That means, any remote/networked solution was more "possible hack" than "designed use". Then I agree to declare this option obsolete. The S.S. section in the SUG can start with an obsolescence note, and the section removed in the SUG for version 0.20.0. (For those who did it before, I think nothing actually changes, and they can continue to use S.S. and just follow older user guides?).
> 
> Regarding Alpaca, if this is not ready for prime time, we can postpone this of course. But you likely know better. You are cordially invited to expand on this later :-)
> 
> If I understand correctly, "ASCOM compliant" still requires XP and Vista support. Or is there a statement from the ASCOM developers that remedies this statement?

@gzotti That's just for the drivers. In the section for Windows components and Programs, only Windows 7 and 10 is required. Also this guideline, as far as I understand it, is just for protecting the brand ASCOM and only applies if you officially want to use their logo and call your software ASCOM compliant. Nevertheless, we should meet those criteria already.

gionkunz commented:
> Just tested on another, stronger notebook. Now I can modify telescope settings, e.g. slew speed, and have to re-connect after this modification. (This is OK, though, as this selection is usually a one-time thing setting).
> However, when I exit Stellarium while connected to the ASCOM simulator telescope, Stellarium crashes. When I disconnect first, Stellarium exits normally. The logfile has no hint. Bug in the simulator, or do real telescopes behave similarly?

@gzotti I could reproduce this issue. The problem was less obvious to me, since this only happened when exiting Stellarium and I didn't check the return code in the console.

The Problem was that somehow to OLE objects are already disposed from the Windows events when the destructor was called. I have now moved the ASCOM OLE object creation completely into the connect and disconnect methods and don't use the destructor anymore. The problem is now fixed and will push that later.

gionkunz commented:
> I finally was able to test under the sky. NEQ3 with ASCOM Skywatcher driver 6.0.6262.15273. Works nicely, but again, Stellarium crashed on closing while I was still connected to the telescope. Maybe there is some timing issue when disabling/freeing resources? I later tried several times again with just the simulator. Sometimes it works, sometimes it crashes when connected and closing the program.
> 
> The telescope config panel again shows ways to set geographical location. Is this standard for all ASCOM telescopes? Then I would really like to see a bidirectional config option in the main Stellarium ASCOM GUI to either set telescope location from Stellarium, or set Stellarium's location from what is set in the telescope.

Yes, ASCOM scopes usually support position, altitude and time to be set. This is done to initialize the pointing model of the scope. I guess this would be a nice feature for some cases, however, given the fact that location and altitude is only required for polar alignment and atmospheric correcting for JNOW inside of the mount, I think it's somewhat neglectable. This is just not part of any routine I can imagine. If people do their polar alignment manually using a polar scope, they anyway use the hand controller and set the location, time and altitude there (if not persisted already). Then they do the polar alignment based on the information from the hand controller and what they observe in the polar scope reticle. Most people I know use an app on their phone for this anyways, or they have other routines.

The only thing I see a benefit, would be if people who are very used to a mount based workflow (hand controller, persistent settings on the mount, GPS on the mount) could use the settings from the mount in Stellarium. However, the benefit is not very big.

Given all those assumptions, I'd rather like to continue with the existing features and push this as is while fixing remaining bugs and documentation. Then, maybe, at a later stage if demands are observed, do any further changes.

gionkunz commented:
> One more question by that earlier ASCOM ignorant (me). "Sync to object" is something that AFAICS our classical control protocols don't support, we always said Stellarium only sends out GOTO commands and users must polar align correctly first. Is it mandatory for ASCOM? Then this is a clear advantage and justifies me switching over :-) For my test, I had set my telescope not very accurately, aligned with 3 stars (result was a few degrees off correct polar alignment), and the drive managed to go to objects that were on the edge of my wide-angle eyepiece. (I know how to improve alignment for photography, but this is not the point here.) On the Stellarium screen the telescope marker also was a bit off the center of the object, but I did not absolutely check the pointing accuracy. When I moved the object into the center of the eyepiece and did "Sync to object", the marker went to the object. However, on next Slew to (next) object, the target was again on the edge of the eyepiece FoV, and marker was (correctly) not on the object.
> I should repeat the tests however, as the ASCOM driver told me I had only a version 2 handcontroller, some things may not work correctly, and I should upgrade to V 3 or 4. That HC says "version 3.35" which should fall into the "V3" classification (?). Not sure if this is responsible for sync not helping to raise accuracy for the next goto.

Almost all ASCOM telescopes support modifying the internal pointing model. Most scope computers store internally a pointing model which is there to correct polar alignment errors, atmospherics refraction errors, time errors etc.  The pointing model is usually constructed during the 1-star, 3-star or whatever star alignment the mount provides. Usually a triangular star alignment provides great results if you navigate inside that triangle. It's a transformation matrix for coordinates which corrects by displacing the coordinates slightly. When your polar alignment is spot on, ~1 arc minute, 1-star alignment, or a pointing model which is transformed by a simple vector, is enough to correct your slews to enough precision. When your polar alignment error is larger and your polar alignment star is near the horizon where you have more atmospheric refraction, the pointing model needs more points to do the correction.

Most mounts will use the sync commands you issue to them with ASCOM as input to their internal pointing model. On a Celestron All-star polar align Hand controller, they will probably just add more stars to the internal algorithm. A scope which only supports 1-star alignment for the pointing model will probably just use every sync you issue to update this one star. This is very vendor specific and I can only make assumptions.

In my setup with my EQ6-R and EQASCOM (which is also your setup with your EQ3), the whole pointing model is off-loaded to EQASCOM and they provide a fantastic customizable algorithm which will use as-many sync points you wish for and switches triangular correction based on your current location.

In my experience, perfect slow commands are rarely a given. I always had problems with my EQ6-R and Stellarium with Stellarium Scope until recently, where I switched the setting in Stellarium Scope and EQASCOM to explicitly use J2000 instead of JNOW. Since then, after issuing 2 or 3 syncs (with my polar alignment close to perfect ~1 arc minutes), were enough to spot on center any star. Also, using the build of my branch here, tested 3 times in real conditions, I was always having great results with slews after syncing to a few stars. I don't sync manually anymore but use plate-solving to do so.

gionkunz commented:
Here's a picture I took from the monkey head nebula last Saturday, running the build here all night long without issues. Slewing was perfect while I showed a few messier objects to a random passenger who was very much interested in what I'm doing there out all alone with all that equipment ;-)

![ngc2174-monkey-head](https://user-images.githubusercontent.com/2203704/68087716-885ec480-fe58-11e9-81ec-dea42fba8a31.jpg)

gionkunz commented:
> Hi, I can confirm this! If I goto a star with the AstroArt Atlas and then solve and sync the image the star is centered in my image but the scope field in Stellarium shows the star near the edge of the scope circle! If I then do a goto from Stellarium the star moves to the edge of my image. I have J now set in AA and Stellarium.
> 
> > One more question by that earlier ASCOM ignorant (me). "Sync to object" is something that AFAICS our classical control protocols don't support, we always said Stellarium only sends out GOTO commands and users must polar align correctly first. Is it mandatory for ASCOM? Then this is a clear advantage and justifies me switching over :-) For my test, I had set my telescope not very accurately, aligned with 3 stars (result was a few degrees off correct polar alignment), and the drive managed to go to objects that were on the edge of my wide-angle eyepiece. (I know how to improve alignment for photography, but this is not the point here.) On the Stellarium screen the telescope marker also was a bit off the center of the object, but I did not absolutely check the pointing accuracy. When I moved the object into the center of the eyepiece and did "Sync to object", the marker went to the object. However, on next Slew to (next) object, the target was again on the edge of the eyepiece FoV, and marker was (correctly) not on the object.
> > I should repeat the tests however, as the ASCOM driver told me I had only a version 2 handcontroller, some things may not work correctly, and I should upgrade to V 3 or 4. That HC says "version 3.35" which should fall into the "V3" classification (?). Not sure if this is responsible for sync not helping to raise accuracy for the next goto.

@iriise What settings do you have in the ASCOM telescope in stellarium? Have you left it to "Let ASCOM device decide" or did you change to "Use Stellarium settings"? Did you try to experiment a bit? What mount do you have? Did you previously use Stellarium Scope and if yes, how did that behave?

iriise commented:
I left it to "let ASCOM device decide" I'll try "Use Stellarium settings" I didn't notice that setting !
I use a Takahashi EM-200 Temma 2 mount with Temma ASCOM driver.
I used Stellarium Scope previously. All GoTo's worked well with SS!
I only experimented with the Jnow and J2000 settings.
Thanks for adding ASCOM to Stellarium!

gionkunz commented:
@alex-w, I see you've already implemented some changes and added the slew method to the TelescopeClient on the master branch. I'm now facing multiple merge conflicts because of this. Do you need that already on the master branch? Couldn't you make this part of this PR? Now, I see you have slight differences for example, I have used `syncTelescopeToSelectedObject` and you use `syncTelescopeWithSelectedObject`. I'm now merging with master and use your changes, however, I think this could have been prevented with communication.

gionkunz commented:
I have fixed the last two remaining issues that I currently know of (closing Stellarium causes segmentation fault when connected to ASCOM and showing the correct type in the type column of the tree view when using an ASCOM scope).

The only missing part is now the documentation 🎉 

@alex-w I saw you've also added a sync button on the master branch, awesome!

Cheers
Gion

gionkunz commented:
@gzotti I've started my work on the LaTeX guides. I tried to preview my changes by compiling but somehow I can't seem to get my changes to show up. Is there anything I need to do in order to preview my changes in Stellarium? Thanks.

alex-w commented:
@gionkunz Hmm... How is more correct for English language: `syncTelescopeToSelectedObject` or `syncTelescopeWithSelectedObject`?

P.S. Could you check unit tests for ASCOM?

P.P.S. In any case sorry for code conflicts

gzotti commented:
> @gzotti I've started my work on the LaTeX guides. I tried to preview my changes by compiling but somehow I can't seem to get my changes to show up. Is there anything I need to do in order to preview my changes in Stellarium? Thanks.

Are you familiar with LaTeX? You need to use pdflatex, or rather use the Makefile (make guide.pdf).
On the Windows Subsystem for Linux + Ubuntu you can install texlive. Please follow instructions in the README.md in the guide subdirectory. 

To see updated PDFs, I recommend SumatraPDF. If you have it open in Adobe Reader, this blocks re-writing a PDF.

But it would be enough if you just write the text. I will likely do some redactional changes later.

gionkunz commented:
@gzotti yes I know LaTeX a bit, however, I was hoping that there's some generation / interpretation task happening that will make the tex files readable directly in Stellarium. I though that's what I see when I go to the Help tab in the Telescope dialog.

gionkunz commented:
"Its command shell is an embarassing relic of the DOS era and a far cry from tools available on the tiniest Linux system" very well put :-D I'm using WSL1 and 2 nowadays, previously I was always working in a Ubuntu VM.

gzotti commented:
No, unfortunately the instructions in the help panel are totally separate from the User Guide. Ah, again, don't forget to exclude the ASCOM part of the instructions in the help panel from non-Windows builds. 

I had been using a Cygwin shell for a decade :-)  I see I could update the README.md to recommend WSL/Ubuntu as primary path, as this just allows following Linux instructions.

sunshuwei commented:
It works perfectly on my  notebook computer.

gionkunz commented:
@gzotti @alex-w I have finally found some time to update the guide in Stellarium and added a section on ASCOM. I have highlighted all the pre-requisites as well as incorporated some notes that should help some users with troubleshooting based on the feedback we had on Cloudy Nights (ASCOM Platform version > 6, 32bit / 64bit issue etc.)

I hope we're now good to go and this update can land in the official beta version and be targeted for 0.19.3 release :-)

gzotti commented on guide/plg_atTheTelescope.tex:
```diff
@@ -837,11 +850,43 @@ \section{INDI}
+the button \button{Choose ASCOM telescope}. The selected device is shown below that button.
+
+By default, Stellarium is trying to ask the ASCOM driver in order to determine if JNow or J2000 coordinates 
+should used for communicating with the mount. If you'd like to override this behavior and tell the ASCOM telescope
```
line 875: should be used

gzotti commented on guide/plg_atTheTelescope.tex:
```diff
@@ -837,11 +850,43 @@ \section{INDI}
 \program{StellariumScope} is a free add-on that enables you to control your telescope with Stellarium. 
 
+Note that Stellarium supports ASCOM telescopes natively since version 0.19.3. There's no need to 
+  use Stellarium Scope in order to use your ASCOM telescopes with Stellarium anymore. We are 
```
I think correct spelling is StellariumScope? (2x)

gionkunz replied:
Fixed

gzotti commented:
Dear @gionkunz , thank you very much, this is easily the TOP (!!) contribution to Stellarium of 2019!

I was following conversations on CN, looked very positive. My own test in late October was mostly perfect, and you fixed that last issue. (Crash when still connected at Stellarium shutdown.)
Code looks very clean as fas as I can tell. 
Also thank you for the (almost) perfect docs in the Guide. If possible please fix the tiny typos before merge, else I will do it.

@alex-w for me it's ready to be merged and go into official beta.

gionkunz commented:
Thank you guys for all your support on this. I have a 10k+ stars open source project myself and I know it's never easy managing your time and priorities with open source. Most people don't understand that, but I really do and I really appreciate how much effort you have put into this project.

I have also tested the build now over many nights and together with the feedback on CloudyNights, we should really be prepared for the official beta.

Thanks again!

albyonegitman commented:
Hi there...
I was wondering if anyone got this ASCOM to work on Linux? I'm on OpenSUSE 15.1 and would be willing to test it if anyone is willing to update it. It's been a while since I've written C/C++ (I currently write bash, Java, Python.... and dare I say it -  C# !! - mainly ASP.NET and MVC Web Apps). I can even take a stab at building it, since I'm currently a Linux Sys Admin by trade.

Thanks!

alex-w commented:
@albyonegitman you can try add ASCOM Alpaca support

mehrdad856 commented:
Hello to all
Can anyone help me with the method of using the Moo Axis command to move the axes of a telescope tropical mount to track a satellite? What is the syntax of the command and how do I define the axes for this command?
Good luck

mehrdad856 commented:
I mean MoveAxis command!

mehrdad856 commented:
I want to move the axes independently and I do not know how to introduce the axes for this command
